pkg/controller/uniteddeployment/allocator.go

```diff
@@ -0,0 +1,256 @@
+	return &replicasAllocator{&n, map[string]int32{}}
+}
+
+type replicasAllocator struct {
+	subsets           *subsetInfos
+	AllocatedReplicas map[string]int32
+}
+
+func (s *replicasAllocator) effectiveReplicasLimitation(replicas int32, subsetReplicasLimits map[string]int32) bool {
+	if subsetReplicasLimits == nil {
+		return false
+	}
+
+	var specifiedReplicas int32
+	for _, replicas := range subsetReplicasLimits {
+		specifiedReplicas += replicas
+	}
+
+	if specifiedReplicas > replicas {
+		return false
+	}
+
+	limitedCount := 0
+	for _, subset := range *s.subsets {
+		if _, exist := subsetReplicasLimits[subset.SubsetName]; exist {
+			limitedCount++
+		}
+	}
+
+	if limitedCount != len(subsetReplicasLimits) {
+		return false
+	}
+
+	return true
+}
+
+func getReplicasAllocator(nameToSubset map[string]*Subset, ud *appsv1alpha1.UnitedDeployment) (map[string]int32, *replicasAllocator) {
+	nameToSubsetDef := map[string]*appsv1alpha1.Subset{}
+	for i := range ud.Spec.Topology.Subsets {
+		nameToSubsetDef[ud.Spec.Topology.Subsets[i].Name] = &ud.Spec.Topology.Subsets[i]
+	}
+
+	ntr := make([]*nameToReplicas, len(nameToSubset))
+	replicaLimits := map[string]int32{}
+
+	idx := 0
+	for name, subset := range nameToSubset {
+		subsetDef, exist := nameToSubsetDef[name]
+		if exist && subsetDef.Replicas != nil {
+			if unitLimit, err := ParseSubsetReplicas(*ud.Spec.Replicas, *subsetDef.Replicas); err == nil {
+				replicaLimits[name] = unitLimit
+			} else {
+				klog.Warningf("Fail to consider the replicas of subset %s/%s when parsing replicaLimits during managing replicas of UnitedDeployment %s/%s: %s",
+					subset.Namespace, subset.Name, ud.Namespace, ud.Name, err)
+			}
+		}
+
+		ntr[idx] = &nameToReplicas{Name: subset.Name, SubsetName: name, Replicas: subset.Spec.Replicas}
+		idx++
+	}
+	replicasAllocator := subsetInfos(ntr).SortToAllocator()
+	return replicaLimits, replicasAllocator
+}
+
+func (s *replicasAllocator) AllocateReplicas(replicas int32, subsetReplicasLimits map[string]int32) {
+	if !s.effectiveReplicasLimitation(replicas, subsetReplicasLimits) {
+		s.normalAllocate(replicas, 0, s.subsets.Len())
+		return
+	}
+
+	var specifiedReplicas int32
+	for _, subset := range *s.subsets {
+		if limit, exist := subsetReplicasLimits[subset.SubsetName]; exist {
+			specifiedReplicas += limit
+			subset.Replicas = limit
+			subset.Specified = true
+		}
+	}
+
+	index := s.subsets.Len() - 1
+	for i := index; i >= 0; i-- {
+		if s.subsets.Get(i).Specified {
+			(*s.subsets)[i], (*s.subsets)[index] = (*s.subsets)[index], (*s.subsets)[i]
+			index--
+		}
+	}
+
+	// apply the replicas to the left subsets
+	s.normalAllocate(replicas, specifiedReplicas, index+1)
+	return
+}
+
+func (s *replicasAllocator) normalAllocate(replicas, excludedReplicas int32, consideredLen int) {
+	var replicasAfterLimit int32
+	for _, nts := range *s.subsets {
+		replicasAfterLimit += nts.Replicas
+	}
+
+	diff := replicas - replicasAfterLimit
+	replicas -= excludedReplicas
```

Comment:
I don't quite understand the entire logic without comments. Can you describe the algorithm?

Reply:
Added some comments.